parachain/primitives/ethereum/src/difficulty.rs

```diff
@@ -0,0 +1,253 @@
+const EXP_DIFFICULTY_PERIOD: u32 = 100000;
+const MINIMUM_DIFFICULTY: u32 = 131072;
+
+#[derive(PartialEq, RuntimeDebug)]
+pub enum BombDelay {
+	// See https://eips.ethereum.org/EIPS/eip-649
+	Byzantium = 3000000,
+	// See https://eips.ethereum.org/EIPS/eip-1234
+	Constantinople = 5000000,
+	// // See https://eips.ethereum.org/EIPS/eip-2384
+	MuirGlacier = 9000000,
+}
+
+/// Describes when hard forks occurred that affect difficulty calculations. These
+/// values are network-specific.
+#[derive(PartialEq, RuntimeDebug)]
+pub struct DifficultyConfig {
+	// Block number on which Byzantium (EIP-649) rules activated
+	pub byzantium_fork_block: u64,
+	// Block number on which Constantinople (EIP-1234) rules activated
+	pub constantinople_fork_block: u64,
+	// Block number on which MuirGlacier (EIP-2384) activated
+	pub muir_glacier_fork_block: u64,
+}
+
+impl DifficultyConfig {
+
+	// Correct block numbers for mainnet and various testnets can be found here:
+	// https://github.com/ethereum/go-ethereum/blob/498458b4102c0d32d7453035a115e6b9df5e485d/params/config.go#L55-L258
+	pub fn mainnet() -> Self {
+		DifficultyConfig {
+			byzantium_fork_block: 4370000,
+			constantinople_fork_block: 7280000,
+			muir_glacier_fork_block: 9200000,
+		}
+	}
+
+	pub fn ropsten() -> Self {
+		DifficultyConfig {
+			byzantium_fork_block: 1700000,
+			constantinople_fork_block: 4230000,
+			muir_glacier_fork_block: 7117117,
+		}
+	}
+
+	pub fn bomb_delay(&self, block_number: u64) -> Option<BombDelay> {
+		if block_number >= self.muir_glacier_fork_block {
+			return Some(BombDelay::MuirGlacier);
+		} else if block_number >= self.constantinople_fork_block {
+			return Some(BombDelay::Constantinople);
+		} else if block_number >= self.byzantium_fork_block {
+			return Some(BombDelay::Byzantium);
+		}
+		None
+	}
+}
+
+/// This difficulty calculation follows Byzantium rules (https://eips.ethereum.org/EIPS/eip-649)
+/// and shouldn't be used to calculate difficulty prior to the Byzantium fork.
+pub fn calc_difficulty(
+	config: &DifficultyConfig,
+	time: u64,
+	parent: &Header,
+) -> Result<U256, &'static str> {
+	let bomb_delay = config.bomb_delay(parent.number + 1)
+		.ok_or("Cannot calculate difficulty for block number prior to Byzantium")?;
+
+	let block_time = time.checked_sub(parent.timestamp).ok_or("Invalid block time")?;
+	let block_time_div_9 = i32::try_from(block_time / 9).or(Err("Invalid block time"))?;
+	let sigma2 = match parent.has_ommers() {
+		true => 2 - block_time_div_9,
+		false => 1 - block_time_div_9,
+	}.max(-99);
+
+	let mut difficulty_without_exp = parent.difficulty;
+	if sigma2 < 0 {
+		difficulty_without_exp -= (parent.difficulty >> DIFFICULTY_BOUND_DIVISOR) * (sigma2 * -1) as u32;
+	} else {
+		difficulty_without_exp += (parent.difficulty >> DIFFICULTY_BOUND_DIVISOR) * sigma2 as u32;
+	}
+
+	difficulty_without_exp = difficulty_without_exp.max(MINIMUM_DIFFICULTY.into());
+
+	let fake_block_number: u32 = u32::try_from(parent.number)
```

Comment:
If `Header::number` is `u64`, why would we want to reduce precision to `u32`?

Reply:
I feel like there was a reason but I can't remember what it was :D I'll keep it u64 and see if I run into an issue again